deployment: build docker image

matz-e opened the pull request:
This should provide an up-to-date docker image to be used in our public-facing CI deployments, i.e., for MVDTool.

See here for a slightly out-of-date configuration: https://github.com/BlueBrain/MVDTool/blob/master/.travis.yml (`--run-tests` → `--test=root` needs to happen, still)

matz-e committed:
deployment: build docker image

matz-e committed:
Use the right agent, attempt 1.3 syntax again.

matz-e committed:
Fixes.

matz-e committed:
Fix dir.

matz-e committed:
Add proxy.

matz-e committed:
Add proxy properly.

matz-e committed:
Lower case env variable.

matz-e committed:
Add https proxy, -sudo

matz-e committed:
Fetch refs right.

matz-e committed:
Remove keyscan.

tristan0x commented on docker/Dockerfile:
```diff
@@ -0,0 +1,79 @@
+RUN spack compiler find && spack compilers
+
+RUN mkdir -p $SPACK_ROOT/etc/spack
+ADD packages.yaml $SPACK_ROOT/etc/spack
```
Here `COPY` is enough. You can also use one single `COPY` instead of severals.

tristan0x commented on docker/Dockerfile:
```diff
@@ -0,0 +1,79 @@
+
+# see this: http://stackoverflow.com/questions/20635472/using-the-run-instruction-in-a-dockerfile-with-source-does-not-work
+RUN rm /bin/sh && ln -s /bin/bash /bin/sh
+RUN . $SPACK_ROOT/share/spack/setup-env.sh
```
Line 59 can be removed right?

matz-e replied:
Sourcing the environment? Don't think so. Re-aliasing the shell could be removed, though.

tristan0x replied:
`source` only applies to the `RUN` instruction. Environment is lost afterward.

matz-e replied:
Like Jenkins… Thanks! I hoped that Docker would be "smarter"

tristan0x commented on docker/Dockerfile:
```diff
@@ -0,0 +1,79 @@
+RUN git fetch
+
+ARG pr
+RUN [[ "x$pr" != "x" ]] && git fetch +refs/pull/$pr/head:refs/remotes/origin/pull/$pr
```
I would declare `ARG pr` before the git clone to make sure to clone if a new PR comes up.

matz-e replied:
I ideally would want to re-use most of the previous build steps to avoid repeating the `apt-get install`. Of course I can move all the git stuff to the bottom of the file.

tristan0x commented on docker/Dockerfile:
```diff
@@ -0,0 +1,79 @@
+ARG pr
+RUN [[ "x$pr" != "x" ]] && git fetch +refs/pull/$pr/head:refs/remotes/origin/pull/$pr
+
+ARG revision=develop
```
Same for revision. IMHO it should be before the git clone

matz-e committed:
Fix git.

matz-e committed:
More fiddling.

matz-e committed:
UPPERCASE

matz-e committed:
.less source

matz-e committed:
explicit /

matz-e committed:
Mess with refspec.

matz-e committed:
More ref magic

matz-e committed:
Remove leftovers.

matz-e committed:
docker: factor out ubuntu configuration, reusable by users.

matz-e committed:
readme: more comments.

matz-e committed:
deploy/ubuntu: add libxml2 to packages.yaml

matz-e committed:
docker: don't push yet.

matz-e commented:
@pramodk any thoughts on this? Since the base image between the Docker ubuntu and our desktops is shared, I extracted some common ground and updated the global README to be workstation-appropriate, too.

matz-e committed:
docker: split build and deployment.

pramodk commented on docker/Dockerfile:
```diff
@@ -0,0 +1,47 @@
+RUN apt-get update
+RUN sed -e 's/#.*//g' ../sysconfig/ubuntu-18.04/packages| \
+        xargs -r apt-get --allow-downgrades \
+                         --allow-remove-essential \
```
I am mssing something obvious : could you remind me how sysconfig dir available before cloning repo?

matz-e replied:
Sorry, that's why the build fails. Haven't fixed that just yet.

pramodk commented on sysconfig/ubuntu-18.04/modules.yaml:
```diff
@@ -0,0 +1,38 @@
+      - neuronmodels
+      - mvapich2
+      - tau
+    blacklist:
```
May be we can add all our packages in bbp applications here?

matz-e replied:
Yes

pramodk commented on sysconfig/ubuntu-18.04/packages.yaml:
```diff
@@ -0,0 +1,107 @@
+    hdf5:
+        paths:
+            hdf5@1.10: /usr/lib/x86_64-linux-gnu/hdf5/serial
+        version: [1.10]
```
I will explicitly add ~mpi for hdf5. Note that our packages are depend on parallel hdf5 now. So could you check parallel hdf5 with corresponding mpi ?

pramodk commented on sysconfig/ubuntu-18.04/packages.yaml:
```diff
@@ -0,0 +1,107 @@
+    all:
+        providers:
+            mpi: [mvapich2]
+            compiler: [gcc]
```
if mvapich2 is default then it will be built from source...I would use OpenMPI here so that it’s easy....also install hdf5-openmpi package for the same

matz-e replied:
OK, will do.

matz-e committed:
docker: fixes

matz-e committed:
config: identical modules for ubuntu and bb5.

matz-e commented:
Ping @pramodk

pramodk reviewed (approved):
Sorry! I thought I did this before
